File: algorithms/284. Peeking Iterator/solution.py
# Below is the interface for Iterator, which is already defined for you.
#
# class Iterator(object):
#     def __init__(self, nums):
#         """
#         Initializes an iterator object to the beginning of a list.
#         :type nums: List[int]
#         """
#
#     def hasNext(self):
#         """
#         Returns true if the iteration has more elements.
#         :rtype: bool
#         """
#
#     def next(self):
#         """
#         Returns the next element in the iteration.
#         :rtype: int
#         """

class PeekingIterator(object):
    def __init__(self, iterator):
        """
        Initialize your data structure here.
        :type iterator: Iterator
        """
        self._it = iterator
        self._data = None
        self.hasNext()

    def peek(self):
        """
        Returns the next element in the iteration without advancing the iterator.
        :rtype: int
        """
        self.hasNext()
        return self._data

    def next(self):
        """
        :rtype: int
        """
        self.hasNext()
        t = self._data
        self._data = None
        return t

    def hasNext(self):
        """
        :rtype: bool
        """
        if self._data is None and self._it.hasNext():
            self._data = self._it.next()
        return self._data is not None

# Your PeekingIterator object will be instantiated and called as such:
# iter = PeekingIterator(Iterator(nums))
# while iter.hasNext():
#     val = iter.peek()   # Get the next element but not advance the iterator.
#     iter.next()         # Should return the same value as [val].

File: algorithms/284. Peeking Iterator/solution.cc
// Below is the interface for Iterator, which is already defined for you.
// **DO NOT** modify the interface for Iterator.
class Iterator {
    struct Data;
	Data* data;
public:
	Iterator(const vector<int>& nums);
	Iterator(const Iterator& iter);
	virtual ~Iterator();
	// Returns the next element in the iteration.
	int next();
	// Returns true if the iteration has more elements.
	bool hasNext() const;
};


class PeekingIterator : public Iterator {
    int data;
    bool hasData;
public:
	PeekingIterator(const vector<int>& nums) : Iterator(nums) {
	    // Initialize any member here.
	    // **DO NOT** save a copy of nums and manipulate it directly.
	    // You should only use the Iterator interface methods.
	    data = (hasData = Iterator::hasNext()) ? Iterator::next() : 0;
	}

    // Returns the next element in the iteration without advancing the iterator.
	int peek() {
        return data;
	}

	// hasNext() and next() should behave the same as in the Iterator interface.
	// Override them if needed.
	int next() {
	    int t = data;
	    data = (hasData = Iterator::hasNext()) ? Iterator::next() : 0;
	    return t;
	}

	bool hasNext() const {
	    return hasData;
	}
};
